## **Reading Food Hygiene Data Analysis**

### Objectives

* Analyse the UK Food Standards Agency dataset, and clean in preparation for data analysis.

### Planned Steps
1. Download and load raw data
2. Filter data to only Reading Local Authority
3. Identify columns needed for analysis - Create new Reading dataset
4. Initial inspection of data
5. Conclusion of data inspection and cleaning strategy
6. Clean data and validate
7. Export cleaned dataset for analysis, visualisation and conclusion

### Input (Data Source)
* **Data Source:** [UK Food Standards Agency - Open Data CSV](https://safhrsprodstorage.blob.core.windows.net/opendatafileblobstorage/FHRS_All_en-GB.csv)

### Outputs
* (3) - Filtered Reading Local Authority dataset
* (7) - Cleaned Reading Local Authority dataset for analysis

---

### Setup
Install and Import required libraries

In [1]:
%pip install pandas numpy

import numpy as np
import pandas as pd

print("Libraries imported successfully!")


[notice] A new release of pip is available: 24.3.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.
Libraries imported successfully!


---
### 1. Download and load raw data.
Raw data downloaded from the Food Standards Agency (link above).

**File Location:** Saved locally in the `data` directory as `FHRS_All_en-GB.csv`

In [2]:
file_path = "../data/FHRS_All_en-GB.csv"  # Set the file path to the CSV file
print(f"Reading data from {file_path}...")  # Print a message indicating the file being read
df_raw = pd.read_csv(file_path, encoding="utf-8-sig", low_memory=False)  # Read the CSV file into a DataFrame
print(f"Data read successfully. Shape of the DataFrame: {df_raw.shape}")  # Print the shape of the DataFrame

Reading data from ../data/FHRS_All_en-GB.csv...
Data read successfully. Shape of the DataFrame: (611521, 23)


---
### 2. Filter data to only Reading Local Authority
- Filter data to Reading
- Look through both outputs to identify any duplicates or things that could become blockers in moving forward.

In [3]:
# Identify the name of the columns to target local authority
df_raw.columns.tolist()  

['AddressLine1',
 'AddressLine2',
 'AddressLine3',
 'AddressLine4',
 'BusinessTypeID',
 'FHRSID',
 'BusinessName',
 'BusinessType',
 'ConfidenceInManagement',
 'Hygiene',
 'Latitude',
 'LocalAuthorityBusinessID',
 'LocalAuthorityCode',
 'LocalAuthorityName',
 'Longitude',
 'NewRatingPending',
 'PostCode',
 'RatingDate',
 'RatingKey',
 'RatingValue',
 'RightToReply',
 'SchemeType',
 'Structural']

In [4]:
# Identify the actual name of the local authority, and sort alphabetically

unique_authorities = df_raw['LocalAuthorityName'].unique().tolist()  # List of unique local authorities
unique_authorities.sort()  # Sort the list alphabetically 
unique_authorities

['Aberdeen City',
 'Aberdeenshire',
 'Adur',
 'Amber Valley',
 'Anglesey',
 'Angus',
 'Antrim and Newtownabbey',
 'Ards and North Down',
 'Argyll and Bute',
 'Armagh City, Banbridge and Craigavon',
 'Arun',
 'Ashfield',
 'Ashford',
 'Babergh',
 'Barking and Dagenham',
 'Barnet',
 'Barnsley',
 'Basildon',
 'Basingstoke and Deane',
 'Bassetlaw',
 'Bath and North East Somerset',
 'Bedford',
 'Belfast City',
 'Bexley',
 'Birmingham',
 'Blaby',
 'Blackburn',
 'Blackpool',
 'Blaenau Gwent',
 'Bolsover',
 'Bolton',
 'Boston',
 'Bournemouth, Christchurch and Poole',
 'Bracknell Forest',
 'Bradford',
 'Braintree',
 'Breckland',
 'Brent',
 'Brentwood',
 'Bridgend',
 'Brighton and Hove',
 'Bristol',
 'Broadland',
 'Bromley',
 'Bromsgrove',
 'Broxbourne',
 'Broxtowe',
 'Buckinghamshire',
 'Burnley',
 'Bury',
 'Caerphilly',
 'Calderdale',
 'Cambridge City',
 'Camden',
 'Cannock Chase',
 'Canterbury City',
 'Cardiff',
 'Carmarthenshire',
 'Castle Point',
 'Causeway Coast and Glens',
 'Central Bedfor

---
 ### 3. Identify columns needed for analysis
We already know the names of the columns from Step 2.

The columns to keep:
| Column Name | What it is | Rationale |
| :--- | :--- | :--- |
| LocalAuthorityBusinessID | Unique reference code given to the establishment | Serves as a unique identifier for the business
| BusinessName | Trading name of the business | Essential for labelling charts and specific lookups
| BusinessType | The category of the business (e.g., Restaurant/Cafe/Retailer) | Crucial for comparative analysis across sectors
| RatingValue | The hygiene score awarded to the business | Primary target metric for distributions and profiling
| RatingDate | The date when the inspection took place | Enables analysis of inspection trends
| AddressLine1 | The first line of the address for the business | Adds granular location context for individual venues 
| Postcode | The postcode for the business | Useful for neighbourhood grouping and mapping
| LocalAuthorityName | The name of the Local Authority (Reading) | Verification anchor confirming filter accuracy

Save new dataset

In [5]:
reading_data = (df_raw[df_raw['LocalAuthorityName'] == 'Reading']  # Filter the DataFrame for Reading local authority
[['LocalAuthorityBusinessID', 'BusinessName', 'BusinessType', 'RatingValue', 'RatingDate', 'AddressLine1', 'PostCode', 'LocalAuthorityName']].copy())  # Select relevant columns

reading_data.to_csv('../data/reading_food_hygiene_data.csv', index=False)  # Save the filtered DataFrame to a new CSV file

display(reading_data.head())  # Display the first few rows of the filtered DataFrame

,LocalAuthorityBusinessID,BusinessName,BusinessType,RatingValue,RatingDate,AddressLine1,PostCode,LocalAuthorityName
413879,137934,2 Chefs Catering Ltd,Other catering premises,5,2024-11-24,NaN,NaN,Reading
413880,80333,3 Love Food,Other catering premises,5,2026-04-17,NaN,NaN,Reading
413881,142094,3B's Cafe,Other catering premises,5,2026-03-17,735 Oxford Road,RG30 1JA,Reading
413882,127516,3Cs Cafe at St Johns Church,Restaurant/Cafe/Canteen,5,2025-09-23,Gosbrook Road,RG4 5AN,Reading
413883,72396,6Pack Supplements,Distributors/Transporters,Exempt,NaN,443 Oxford Road,RG30 1HD,Reading


---
### Dataset Optimisation
By filtering the nationwide UK Food Standards Agency raw dataset **(140MB)**, to our target area of **Reading** and selecting only the 8 target columns, the processed dataset has been streamlined to just **132KB**.

**Why this matters:** This drastically speeds up execution times for visualisation tools, while keeping our project lightweight. 

**File Location:** Saved locally in the `data` directory as `reading_food_hygiene_data.csv`

As this file is significantly smaller, it then gave me a chance to open the file in Microsoft Excel, to inspect the file for missing data and any significant problems before moving any further.

---
### 4. Initial inspection of data
- Load `reading_food_hygiene_data.csv`
- Check the initial shape of the dataset
- Check the columns, data type and non-null counts
- Check for missing values across all columns
- Check the percentage of missing data for each column
- Check the summary statistics for each column
- Consider missing data and how to resolve it

In [6]:
file_path = "../data/reading_food_hygiene_data.csv"  # Set the file path to the new CSV file
df = pd.read_csv(file_path)  # Read the CSV file into a DataFrame

print(f'Shape:\n{df.shape}')  # Print the initial shape of the dataframe

Shape:
(1508, 8)


In [7]:
print(df.dtypes)  # Print columns and data types of each

LocalAuthorityBusinessID    str
BusinessName                str
BusinessType                str
RatingValue                 str
RatingDate                  str
AddressLine1                str
PostCode                    str
LocalAuthorityName          str
dtype: object


In [8]:
df.isnull().sum()  # Print the number of null values in each column

LocalAuthorityBusinessID      0
BusinessName                  0
BusinessType                  0
RatingValue                   0
RatingDate                  188
AddressLine1                158
PostCode                    187
LocalAuthorityName            0
dtype: int64

In [9]:
df.isnull().sum() / len(df) * 100  # Print the percentage of null values in each column

LocalAuthorityBusinessID     0.000000
BusinessName                 0.000000
BusinessType                 0.000000
RatingValue                  0.000000
RatingDate                  12.466844
AddressLine1                10.477454
PostCode                    12.400531
LocalAuthorityName           0.000000
dtype: float64

In [10]:
df.describe(include='all') # Check summary statistics for all columns

,LocalAuthorityBusinessID,BusinessName,BusinessType,RatingValue,RatingDate,AddressLine1,PostCode,LocalAuthorityName
count,1508,1508,1508,1508,1320,1350,1321,1508
unique,1508,1422,13,8,553,1121,590,1
top,137934,Tesco,Restaurant/Cafe/Canteen,5,2026-03-11,Bridge Street,RG1 2AG,Reading
freq,1,10,334,853,10,37,28,1508


---
### Missing Data
All columns except three contain data

By the above, we can see that we are missing values in the following columns:
- **188** in RatingDate - These equate to 12.46%
- **158** in AddressLine1 - These equate to 10.47%
- **187** in PostCode - These equate to 12.40%

We now need to look at these columns to check what the missing data is, and make a decision on what to do with them

In [11]:
# Filter to the rows where any columns have null values
missing_rows = df[df['RatingDate'].isnull() | df['AddressLine1'].isnull() | df['PostCode'].isnull()]
missing_rows


,LocalAuthorityBusinessID,BusinessName,BusinessType,RatingValue,RatingDate,AddressLine1,PostCode,LocalAuthorityName
0,137934,2 Chefs Catering Ltd,Other catering premises,5,2024-11-24,NaN,NaN,Reading
1,80333,3 Love Food,Other catering premises,5,2026-04-17,NaN,NaN,Reading
4,72396,6Pack Supplements,Distributors/Transporters,Exempt,NaN,443 Oxford Road,RG30 1HD,Reading
12,99401,A & G Retail,Retailers - other,5,2024-06-03,NaN,NaN,Reading
13,144943,A J Bakes & More,Other catering premises,4,2025-05-08,NaN,NaN,Reading
...,...,...,...,...,...,...,...,...
1473,06916,Woodley Grill,Mobile caterer,AwaitingInspection,NaN,NaN,NaN,Reading
1480,06793,Yan's Kitchen,Other catering premises,AwaitingInspection,NaN,NaN,NaN,Reading
1482,140746,Yattendon Honey,Other catering premises,5,2025-09-09,NaN,NaN,Reading
1501,06962,Zeus Greek Gyros,Restaurant/Cafe/Canteen,AwaitingInspection,NaN,2 Bridge Street,RG1 2LU,Reading


**Initial Thoughts - Missing RatingDate**
- The businesses that are missing ***RatingDate*** values appear to be due to ***AwaitingInspection*** or ***Exempt***.
  - Check to see if the value of the missing *RatingDate* correlates to *AwaitingInspection* or *Exempt*.
  - Re-check if RatingDate is then zero afterwards.
  - **Consider removal** - as even though they equate for 12.46%, they serve no value to answering our questions or helping to visualise the dataset.

In [12]:
# Check which businesses have a missing RatingDate, against the rating values (AwaitingInspection and Exempt).
unrated_or_exempt = df[(df['RatingDate'].isnull()) & (df['RatingValue'].isin(['AwaitingInspection', 'Exempt']))]
unrated_or_exempt

,LocalAuthorityBusinessID,BusinessName,BusinessType,RatingValue,RatingDate,AddressLine1,PostCode,LocalAuthorityName
4,72396,6Pack Supplements,Distributors/Transporters,Exempt,NaN,443 Oxford Road,RG30 1HD,Reading
21,144194,Ace Catering,Mobile caterer,AwaitingInspection,NaN,87 Southcote Lane,RG30 3AH,Reading
22,33111,Admiral Casino,Retailers - other,Exempt,NaN,93-98 Friar Street,RG1 1EP,Reading
25,130668,Age UK,Retailers - other,Exempt,NaN,20a School Road,RG31 5AN,Reading
27,28001,Ahmed Chem Ltd,Retailers - other,Exempt,NaN,71 Basingstoke Road,RG2 0ER,Reading
...,...,...,...,...,...,...,...,...
1460,58646,Wickes Building Supplies,Retailers - other,Exempt,NaN,709-731 Oxford Road,RG30 1PR,Reading
1467,06939,WithBrand,Distributors/Transporters,Exempt,NaN,450 Brook Green Drive,RG2 6UU,Reading
1473,06916,Woodley Grill,Mobile caterer,AwaitingInspection,NaN,NaN,NaN,Reading
1480,06793,Yan's Kitchen,Other catering premises,AwaitingInspection,NaN,NaN,NaN,Reading


In [13]:
# Count the remaining null values in the DataFrame
remaining_nulls = df.drop(unrated_or_exempt.index).isnull().sum()
remaining_nulls

LocalAuthorityBusinessID      0
BusinessName                  0
BusinessType                  0
RatingValue                   0
RatingDate                    0
AddressLine1                112
PostCode                    141
LocalAuthorityName            0
dtype: int64

**Initial Thoughts - AddressLine1 and PostCode**
- Is there a connection between the type of Business, and addresses and postcodes which would explain why this data is missing?
  - Check to see if the value of the missing *AddressLine1* and *PostCode* correlate to a specific *BusinessType*
  - Is there a specific reason? (Home-based/Mobile food businesses - GDPR perhaps?)
  - **Consider imputing missing values with *'Unknown'*** - This would work for both *AddressLine1* and *PostCode*, as regardless of the missing data, the remaining that is populated would help support answering our **Key Analytical Questions**, and visualisations.
  - Consider adding an additional Postcode District (e.g., RG1, RG2), to enable regional visualisations.

In [14]:
# Count of missing address and postcodes grouped by business type
missing_address = df[df['AddressLine1'].isnull() | df['PostCode'].isnull()]['BusinessType'].value_counts()
missing_address

BusinessType
Other catering premises                  124
Mobile caterer                            17
Restaurant/Cafe/Canteen                   15
Retailers - other                         14
Takeaway/sandwich shop                     7
Hospitals/Childcare/Caring Premises        6
Manufacturers/packers                      4
Pub/bar/nightclub                          2
Distributors/Transporters                  1
Retailers - supermarkets/hypermarkets      1
Name: count, dtype: int64

In [15]:
# Look at breakdown of missing addresses, to consider business names
pd.set_option('display.max_rows', None)
missing_businesses = df[df['AddressLine1'].isnull() | df['PostCode'].isnull()][['BusinessName', 'BusinessType', 'AddressLine1', 'PostCode']]
missing_businesses

,BusinessName,BusinessType,AddressLine1,PostCode
0,2 Chefs Catering Ltd,Other catering premises,NaN,NaN
1,3 Love Food,Other catering premises,NaN,NaN
12,A & G Retail,Retailers - other,NaN,NaN
13,A J Bakes & More,Other catering premises,NaN,NaN
17,Abikwa Bakes,Other catering premises,NaN,NaN
29,Ai La Home Cooking,Other catering premises,NaN,NaN
33,Albie's Bakes,Other catering premises,NaN,NaN
43,Amanda.jay _art,Other catering premises,NaN,NaN
51,Amicus Residential Care,Other catering premises,NaN,NaN
52,Amicus Residential Care Home,Hospitals/Childcare/Caring Premises,NaN,NaN


---
### 5. Conclusion of data inspection and cleaning strategy

Based on the initial inspection of the dataset, the following observations and data gaps were identified, leading to this cleaning strategy.

* **Missing / Unrated Data:** Rows missing a rating or inspection date will be dropped, as they lack the scoring information required for analysis.

* **Missing Addresses and Postcodes:** Null values in `AddressLine1` and `Postcode` will be imputed with `Unknown` to preserve the valid hygiene score records.

* **Postcode Feature Engineering:** Full postcodes will be split to extract the district within Reading (e.g., `RG1`, `RG2`) into a new `PostCodeDistrict` column to enable cleaner regional visualisations.

* **Data Type Conversions:** Hygiene ratings will be converted from strings to integers to allow for proper numerical calculations and visualisations.

* **Date formatting:** The `RatingDate` column will be changed to a standard date format to support tracking trends over time.

---
### 6. Clean data and validate

Following the cleaning strategy above, we will now apply the cleaning steps.

In [16]:
# Create a working copy for cleaning
df_rfh_clean = df.copy()

**Missing / Unrated Data:** Rows missing a rating or inspection date will be dropped, as they lack the scoring information required for analysis.

In [17]:
# Check shape before cleaning
print(df_rfh_clean.shape)

# Drop rows missing a rating or inspection date
df_rfh_clean = df_rfh_clean.dropna(subset=['RatingValue', 'RatingDate'])

# Validation - Check removal of rows
print(df_rfh_clean.shape)

(1508, 8)
(1320, 8)


**Missing Addresses and Postcodes:** Null values in `AddressLine1` and `Postcode` will be imputed with `Unknown` to preserve the valid hygiene score records.

In [18]:
# Impute missing values in AdressLine1 and Postcode with 'Unknown'
df_rfh_clean['AddressLine1'] = df_rfh_clean['AddressLine1'].fillna('Unknown')
df_rfh_clean['PostCode'] = df_rfh_clean['PostCode'].fillna('Unknown')

# Validation - Check remaining null values
print(df_rfh_clean.isnull().sum())  # Print the number of null values in each column

LocalAuthorityBusinessID    0
BusinessName                0
BusinessType                0
RatingValue                 0
RatingDate                  0
AddressLine1                0
PostCode                    0
LocalAuthorityName          0
dtype: int64


**Postcode Feature Engineering:** Full postcodes will be split to extract the district within Reading (e.g., `RG1`, `RG2`) into a new `PostCodeDistrict` column to enable cleaner regional visualisations.

In [19]:
# Split postcodes to give a new column of PostCodeDistrict
df_rfh_clean['PostCodeDistrict'] = df_rfh_clean['PostCode'].apply(lambda x: str(x).split(' ')[0] if pd.notnull(x) else 'Unknown')

# Validate the addition
df_rfh_clean.head()

,LocalAuthorityBusinessID,BusinessName,BusinessType,RatingValue,RatingDate,AddressLine1,PostCode,LocalAuthorityName,PostCodeDistrict
0,137934,2 Chefs Catering Ltd,Other catering premises,5,2024-11-24,Unknown,Unknown,Reading,Unknown
1,80333,3 Love Food,Other catering premises,5,2026-04-17,Unknown,Unknown,Reading,Unknown
2,142094,3B's Cafe,Other catering premises,5,2026-03-17,735 Oxford Road,RG30 1JA,Reading,RG30
3,127516,3Cs Cafe at St Johns Church,Restaurant/Cafe/Canteen,5,2025-09-23,Gosbrook Road,RG4 5AN,Reading,RG4
5,140230,7 Day European Supermarket,Retailers - other,4,2024-12-07,181 Oxford Road,RG1 7UZ,Reading,RG1


**Data Type Conversions:** Hygiene ratings will be converted from strings to integers to allow for proper numerical calculations and visualisations.

In [20]:
df_rfh_clean['RatingValue'] = df_rfh_clean['RatingValue'].astype(int)

# Validate the data type conversion
df_rfh_clean['RatingValue'].dtype

dtype('int64')

**Date formatting:** The `RatingDate` column will be changed to a standard date format to support tracking trends over time.

In [21]:
# Convert RatingDate to datetime format
df_rfh_clean['RatingDate'] = pd.to_datetime(df_rfh_clean['RatingDate'])

# Validate the conversion
df_rfh_clean['RatingDate'].dtype

dtype('<M8[us]')

---
### Validation
Now the data cleaning steps have been completed. I now need to carry out a final validation check.

This ensures that the cleaned dataset is correctly structured, and reliable before moving forward to data analysis and visualisations.

In [22]:
# Check the rows and number of columns
df_rfh_clean.shape

(1320, 9)

In [23]:
# Check the data types of each column
df_rfh_clean.dtypes

LocalAuthorityBusinessID               str
BusinessName                           str
BusinessType                           str
RatingValue                          int64
RatingDate                  datetime64[us]
AddressLine1                           str
PostCode                               str
LocalAuthorityName                     str
PostCodeDistrict                       str
dtype: object

In [24]:
# Check the number of null counts in each column
df_rfh_clean.isnull().sum()

LocalAuthorityBusinessID    0
BusinessName                0
BusinessType                0
RatingValue                 0
RatingDate                  0
AddressLine1                0
PostCode                    0
LocalAuthorityName          0
PostCodeDistrict            0
dtype: int64

The validation checks confirm that the cleaning strategy has been carried out successfully on `df_rfh_clean`. 

The missing values have been handled, data types (including datetime formatting) are correctly structured, and the new postcode district column has been created.

---
### 7. Export cleaned dataset for analysis, visualisation and conclusion

We will now save our final cleaned dataset (`df_rfh_clean`) to a CSV file called `rfh_clean.csv` within the `data/` directory, so it can easily be loaded for exploratory data analysis and visualisations.

In [26]:
# Export the cleaned dataset to the data directory
df_rfh_clean.to_csv('../data/rfh_clean.csv', index=False)
print("Cleaned dataset exported successfully to '../data/rfh_clean.csv'")

Cleaned dataset exported successfully to '../data/rfh_clean.csv'


The dataset has now been successfully exported and some of the names appear to be long. If they turn into problems at a later stage this will be addressed on a case by case basis, instead of cleaning them at this stage. This is done to preserve the integrity of the raw data.